### PR TITLE
Add HTTPInvocationClient, a wrapper that encompasses an HTTPOperation…

### PR DESCRIPTION
**Pull request by pbthif**
…s client with the invocation context and retry strategy

*Issue #, if available:*

*Description of changes:*

The main goal of this change is to add `HTTPInvocationClient`, which is an utility wrapper over a `HTTPOperationsClient` object. The invocation client is encompasses the retry strategy and invocation context, which are stored at initialization and used for each request.

Ancillary changes:
- Add a default `HTTPClientDelegate` for JSON requests/responses.
- Add a mock version of `HTTPInvocationClient` so that it can be used in unit tests.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Sources/SmokeHTTPClient/HTTPInvocationClient.swift [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+//  HTTPInvocationClient.swift
+//  SmokeHTTPClient
+//
+
+import Foundation
+import AsyncHTTPClient
+import NIOHTTP1
+
+public protocol HTTPInvocationClientProtocol {
+     func shutdown() async throws
```
Comment (tachyonics) [inline Sources/SmokeHTTPClient/HTTPInvocationClient.swift:23]: Nickpick: indentation

File: Sources/SmokeHTTPClient/HTTPInvocationClient.swift [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        }
+    }
+
+    public func executeRetriableWithoutOutput<InputType: HTTPRequestInputProtocol>(
+        endpoint: URL? = nil,
+        endpointPath: String,
+        httpMethod: HTTPMethod,
+        operation: String? = nil,
+        input: InputType) async throws {
+            guard endpoint != nil || self.httpClient.endpointHostName != "" else {
```
Comment (tachyonics) [inline Sources/SmokeHTTPClient/HTTPInvocationClient.swift:221]: I would include a comment just to make it clear what conceptually we are detecting here.

File: Sources/SmokeHTTPClient/HTTPInvocationClient.swift (hunk 1 above)
Comment (pbthif) [reply Sources/SmokeHTTPClient/HTTPInvocationClient.swift:23]: Fixed.

File: Sources/SmokeHTTPClient/HTTPInvocationClient.swift (hunk 2 above)
Comment (pbthif) [reply Sources/SmokeHTTPClient/HTTPInvocationClient.swift:221]: Added this here and in other similar places:
```
// Users can either specify a default endpoint for the `HTTPInvocationClient` instance
// or pass in an request-specific endpoint. Doing neither is an error.
```

**Review by tachyonics — APPROVED** [review]
LGTM, thanks!

**Pull request merged**